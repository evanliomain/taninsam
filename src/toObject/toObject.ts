/**
 * @module array=>object
 */
/**
 * Convert an array to an associative map store in a javascript object.
 * @param keyGetter A function to extract the key of the association.
 * @param valueGetter A function to extract the value of the association.
 * @return the function to apply on the array to do convert it
 * into a javascript object
 * @example
 * ```
 * toObject<(string | number)[][]>(
 *  item => item[0], item => item[1])([['a', 1], ['b', 'c']]
 * ) // {a: 1, b: 'c'}
 * ```
 * @example Using the chain
 * ```
 * chain([['a', 1], ['b', 'c']])
 *   .chain(toObject<(string | number)[][]>(item => item[0], item => item[1]))
 *   .value() // {a: 1, b: 'c'}
 * ```
 */
export function toObject<T, U>(
  keyGetter: (x: T) => string,
  valueGetter: (x: T) => U
): (array: ReadonlyArray<T>) => { [key: string]: U } {
  return (array: ReadonlyArray<T>) =>
    array.reduce(
      (mapObject: { [key: string]: U }, element: T): { [key: string]: U } => ({
        ...mapObject,
        [keyGetter(element)]: valueGetter(element)
      }),
      {}
    );
}
