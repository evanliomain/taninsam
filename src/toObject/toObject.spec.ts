import { toObject } from './toObject';

describe('toObject function', () => {
  const input = [['a', 1], ['b', 'c']];

  test('is a pure function', () => {
    toObject(x => `${x}`, x => x)(input);
    expect(input).toMatchSnapshot();
  });

  test("[['a', 1], ['b', 'c']] |> toObject(key => key, value => value + value) === {a: 1, b: 'c'}", () => {
    expect(toObject(item => item[0], item => item[1])(input)).toEqual({
      a: 1,
      b: 'c'
    });
  });
});
