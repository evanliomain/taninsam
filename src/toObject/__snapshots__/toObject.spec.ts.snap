// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toObject function [['a', 1], ['b', 'c']] |> toObject(item => item[0], item => item[1]) === {a: 1, b: 'c'} 1`] = `
Object {
  "a": 1,
  "b": "c",
}
`;

exports[`toObject function is a pure function 1`] = `
Array [
  Array [
    "a",
    1,
  ],
  Array [
    "b",
    "c",
  ],
]
`;
